rust/lance/src/datafusion/logical_expr.rs

```diff
@@ -34,6 +38,45 @@ fn resolve_value(expr: &Expr, data_type: &DataType) -> Result<Expr> {
     }
 }
 
+/// A simple helper function that interprets an Expr as a string scalar
+/// or returns None if it is not.
+pub fn get_as_string_scalar_opt(expr: &Expr) -> Option<&String> {
+    match expr {
+        Expr::Literal(ScalarValue::Utf8(Some(s))) => Some(s),
+        _ => None,
+    }
+}
+
+// As part of the DF 37 release there are now two different ways to
+// represent a nested field access in `Expr`.  The old way is to use
+// `Expr::field` which returns a `GetStructField` and the new way is
+// to use `Expr::ScalarFunction` with a `GetFieldFunc` UDF.
+//
+// Currently, the old path leads to bugs in DF.  This is probably a
+// bug and will probably be fixed in a future version.  In the meantime
+// we need to make sure we are always using the new way to avoid this
+// bug.  This trait adds field_newstyle which lets us easily create
+// logical `Expr` that use the new style.
+pub trait ExprExt {
+    // Helper function to replace Expr::field in DF 37 since DF
+    // confuses itself with the GetStructField returned by Expr::field
+    fn field_newstyle(&self, name: &str) -> Expr;
```

Comment:
This was the most complicated change.  The comment explains it ok.  I expect this may eventually go away (e.g. when https://github.com/apache/datafusion/issues/10181 is resolved)